pkg/cmd/roachtest/roachtestutil/mixedversion/mutators.go

```diff
@@ -403,26 +403,50 @@ func (m panicNodeMutator) Generate(
 	rng *rand.Rand, plan *TestPlan, planner *testPlanner,
 ) []mutation {
 	var mutations []mutation
-	maxPanics := len(plan.upgrades)
-	numPanics := 1 + rng.Intn(maxPanics)
-
+	upgrades := randomUpgrades(rng, plan)
 	idx := newStepIndex(plan)
-	possiblePointsInTime := plan.
-		newStepSelector().
-		// We don't want to panic concurrently with other steps, and inserting before a concurrent step
-		// causes the step to run concurrently with that step, so we filter out any concurrent steps.
-		Filter(func(s *singleStep) bool {
-			return s.context.System.Stage >= InitUpgradeStage && !idx.IsConcurrent(s)
-		})
+	nodeList := planner.currentContext.System.Descriptor.Nodes
+
+	for _, upgrade := range upgrades {
+		possiblePointsInTime := upgrade.
+			// We don't want to panic concurrently with other steps, and inserting before a concurrent step
+			// causes the step to run concurrently with that step, so we filter out any concurrent steps.
+			Filter(func(s *singleStep) bool {
+				return s.context.System.Stage >= InitUpgradeStage && !idx.IsConcurrent(s)
+			})
 
-	for range numPanics {
-		nodeList := planner.currentContext.System.Descriptor.Nodes
 		targetNode := nodeList.SeededRandNode(rng)
-		addRandomly := possiblePointsInTime.
+		stepToPanic := possiblePointsInTime.RandomStep(rng)
+
+		isInvalidStartStep := func(s *singleStep) bool {
+			// Restarting the system on a node while a node is already dead can
+			// cause the cluster to lose quorum, so we avoid any system starts.
+			_, validStartSystem := s.impl.(restartWithNewBinaryStep)
+			// Waiting for stable cluster version targets every node in
+			// the cluster, so a node cannot be dead during this step.
+			_, waitForStable := s.impl.(waitForStableClusterVersionStep)
+			// Many hook steps do not support running with a dead node,
+			// so we avoid inserting after a hook step.
+			_, runHook := s.impl.(runHookStep)
+
+			return validStartSystem || waitForStable || runHook
+		}
+		_, afterPanicSteps := possiblePointsInTime.CutAfter(func(s *singleStep) bool {
```

Comment:
nit: lets "explain in english" what we're trying to do here. e.g: We want to restart the node after we've panicked it but before we attempt to run a step that in incompatible with unavailable nodes.